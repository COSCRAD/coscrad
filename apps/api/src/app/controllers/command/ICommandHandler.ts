import { ICommand } from './ICommand';

export const Ack = 'acknowledgement';

export interface ICommandHandler {
    execute(command: ICommand): Promise<string>;
}
