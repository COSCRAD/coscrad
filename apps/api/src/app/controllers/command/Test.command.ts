import { ApiProperty } from '@nestjs/swagger';
import { ICommand } from './ICommand';

export class TestCommand implements ICommand {
    type = 'TestCommand';

    @ApiProperty()
    foo: number;

    @ApiProperty()
    bar: string;
}
