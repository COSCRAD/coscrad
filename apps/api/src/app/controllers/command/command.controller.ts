import { AGGREGATE_COMPOSITE_IDENTIFIER } from '@coscrad/api-interfaces';
import { Ack, CommandHandlerService } from '@coscrad/commands';
import {
    Body,
    Controller,
    MessageEvent,
    Post,
    Request,
    Res,
    Sse,
    UnauthorizedException,
    UseFilters,
    UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { Observable, Subject } from 'rxjs';
import { CoscradUserWithGroups } from '../../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import httpStatusCodes from '../../constants/httpStatusCodes';
import { CommandWithGivenTypeNotFoundExceptionFilter } from '../exception-handling/exception-filters/command-with-given-type-not-found.filter';
import { NoCommandHandlerForCommandTypeFilter } from '../exception-handling/exception-filters/no-command-handler-for-command-type.filter';
import sendInternalResultAsHttpResponse from '../resources/common/sendInternalResultAsHttpResponse';
import { CommandFSA } from './command-fsa/command-fsa.entity';

export const AdminJwtGuard = AuthGuard('jwt');

@ApiTags('commands')
@Controller('commands')
/**
 * TODO [https://www.pivotaltracker.com/story/show/182785593]
 * Enforce RBAC for commands in a guard isntead.
 */
/**
 * The next two filters convert a thrown error to a returned error (400) when an
 * invalid command type is provided by the user.
 *
 * TODO [https://www.pivotaltracker.com/story/show/182785593]
 * We may want to do this in a pipe in the future.
 */
@UseFilters(new CommandWithGivenTypeNotFoundExceptionFilter())
@UseFilters(new NoCommandHandlerForCommandTypeFilter())
export class CommandController {
    private readonly commandResultSubject = new Subject<MessageEvent>();

    constructor(private readonly commandHandlerService: CommandHandlerService) {}

    @ApiBearerAuth('JWT')
    @UseGuards(AdminJwtGuard)
    @Post('')
    async executeCommand(@Request() req, @Res() res, @Body() commandFSA: CommandFSA) {
        const { user } = req;

        if (!user || !(user instanceof CoscradUserWithGroups)) {
            throw new UnauthorizedException();
        }

        if (!user.isAdmin()) {
            throw new UnauthorizedException();
        }

        /**
         * Note that we defer command type validation to the command handler.
         * This is because we want to keep the controller free of domain
         * logic. If we want to drive commands via a CLI, it shouldn't need
         * to know about http.
         */
        const { type, payload } = commandFSA;

        const result = await this.commandHandlerService.execute(
            { type, payload },
            { userId: user.id }
        );

        if (result !== Ack) return sendInternalResultAsHttpResponse(res, result);

        this.commandResultSubject.next({
            data: { aggregateCompositeIdentifier: payload[AGGREGATE_COMPOSITE_IDENTIFIER] },
        });

        return res.status(httpStatusCodes.ok).send();
    }

    @Sse('notifications')
    commandSuccessNotifications(): Observable<MessageEvent> {
        return this.commandResultSubject.asObservable();
    }
}
