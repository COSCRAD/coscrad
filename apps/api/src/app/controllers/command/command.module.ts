import { Module, OnApplicationBootstrap } from '@nestjs/common';
import { CommandFinderService } from './CommandFinder.service';
import CommandHandlerService from './CommandHandler.service';

@Module({
    providers: [CommandHandlerService, CommandFinderService],
    exports: [CommandHandlerService],
})
export class CommandModule implements OnApplicationBootstrap {
    constructor(
        private readonly finderService: CommandFinderService,
        private readonly commandHanlderService: CommandHandlerService
    ) {}

    onApplicationBootstrap() {
        const commandHandlers = this.finderService.explore();

        commandHandlers.forEach((handler) => this.commandHanlderService.registerHandler(handler));

        console.log({
            commandHandlerService: JSON.stringify(this.commandHanlderService),
        });
    }
}
