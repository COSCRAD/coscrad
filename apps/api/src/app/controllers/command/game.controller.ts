import { Controller, Get, Param } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AggregateId } from '../../../domain/types/AggregateId';
import { isNullOrUndefined } from '../../../domain/utilities/validation/is-null-or-undefined';
import { NotFound } from '../../../lib/types/not-found';
import { ArangoDatabaseForCollection } from '../../../persistence/database/arango-database-for-collection';
import { ArangoCollectionId } from '../../../persistence/database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../../../persistence/database/database.provider';

type GameDto = {
    id: AggregateId;
    name: string;
};

/**
 * Note that this is a temporary work-around for serving up legacy data.
 * In the future we would have dedicated domain and view models or possibly
 * a separate back-end for game data.
 */
@Controller('games')
export class GameController {
    private readonly gamesDatabase: ArangoDatabaseForCollection<GameDto>;

    constructor(
        private readonly configService: ConfigService,
        databaseProvider: ArangoDatabaseProvider
    ) {
        this.gamesDatabase = databaseProvider.getDatabaseForCollection(
            'games' as ArangoCollectionId
        );
    }

    @Get(':name')
    async fetchGame(@Param('name') nameToFind: string) {
        // Note that the config is technically a string 'false', we should deal with this later
        if (this.configService.get<string>('SHOULD_ENABLE_LEGACY_GAMES_ENDPOINT') === 'false')
            // TODO Send correct error code
            return 'Not Available';

        const allGames = await this.gamesDatabase.fetchMany();

        const searchResult = allGames.find(({ name }) => nameToFind === name);

        if (isNullOrUndefined(searchResult)) return NotFound;

        // We are skipping any  kind of view layer here
        return searchResult;
    }
}
