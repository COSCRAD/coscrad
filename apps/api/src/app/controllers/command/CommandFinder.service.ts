import { Injectable } from '@nestjs/common';
import { ModulesContainer } from '@nestjs/core';
import { COMMAND_HANDLER_METADATA } from './decorators/CommandHandler.decorator';
import { ICommandHandler } from './ICommandHandler';

@Injectable()
export class CommandFinderService {
    constructor(private readonly modulesContainer: ModulesContainer) {}

    explore(): ICommandHandler[] {
        const modules = [...this.modulesContainer.values()];

        const result = modules
            .map((module) => [...module.providers.values()])
            .reduce((a, b) => a.concat(b), [])
            .filter((element) => !!element)
            .map(({ instance }) => instance)
            .filter((instance) => !!instance)
            .filter((instance) => !!instance.constructor)
            .map((instance) => instance.constructor)
            .map((constructor) => Reflect.getMetadata(COMMAND_HANDLER_METADATA, constructor))
            .filter((meta) => !!meta);

        console.log({
            foundCommandHandlers: result,
        });

        return result;
    }
}
