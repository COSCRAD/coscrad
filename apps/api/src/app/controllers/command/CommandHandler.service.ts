import { InternalError } from '../../../lib/errors/InternalError';
import { ResultOrError } from '../../../types/ResultOrError';
import { ICommand } from './ICommand';
import { ICommandHandler } from './ICommandHandler';

export default class CommandHandlerService {
    #handlers: Map<string, ICommandHandler> = new Map();

    registerHandler(type, handler: ICommandHandler) {
        console.log(`Registering command handler with type:${type}, handler: ${handler}`);

        // @ts-expect-error fix types soon
        this.#handlers.set(type, new handler());
    }

    async execute(command: ICommand): Promise<ResultOrError<string>> {
        const { type } = command;

        const handler = this.#handlers.get(type);

        if (!handler)
            throw new InternalError(`There is no handler registered for the command: ${type}`);

        return handler.execute(command);
    }
}
