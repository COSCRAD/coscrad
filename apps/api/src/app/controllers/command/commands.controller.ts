import { Body, Controller, Post, Query, Res } from '@nestjs/common';
import { ApiOkResponse, ApiTags } from '@nestjs/swagger';
import { plainToInstance } from 'class-transformer';
import isStringWithNonzeroLength from '../../../lib/utilities/isStringWithNonzeroLength';
import httpStatusCodes from '../../constants/httpStatusCodes';
import CommandHandlerService from './CommandHandler.service';
import { TestCommand } from './Test.command';

@ApiTags('commands')
@Controller('commands')
export class CommandController {
    constructor(private readonly commandHandlerService: CommandHandlerService) {}

    @ApiOkResponse()
    @Post('')
    async issueCommand(@Res() res, @Query('type') type: string, @Body() payload: TestCommand) {
        if (!isStringWithNonzeroLength(type))
            return res
                .status(httpStatusCodes.badRequest)
                .send(`Invalid command type: ${JSON.stringify(type)}`);

        console.log(`You issued the command: ${type}`);

        const command = plainToInstance(TestCommand, payload);

        console.log({
            commandPayload: command,
        });

        const result = await this.commandHandlerService.execute(
            plainToInstance(TestCommand, payload)
        );

        if (result === 'ack') return res.status(httpStatusCodes.ok).send();

        res.status(httpStatusCodes.internalError).send(`Oopsie!`);
    }
}
