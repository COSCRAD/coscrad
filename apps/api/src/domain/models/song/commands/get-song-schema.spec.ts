import { getCoscradDataSchema } from '@coscrad/data-types';
import { AddSong } from './add-song.command';

describe('when getting the schema for ADD_SONG command', () => {
    it('should return the correct value', () => {
        const schema = getCoscradDataSchema(AddSong);

        expect(schema).toBeTruthy();

        expect(schema).not.toEqual({});

        expect(schema).toMatchSnapshot();
    });
});
