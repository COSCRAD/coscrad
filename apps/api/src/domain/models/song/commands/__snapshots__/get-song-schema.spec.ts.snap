// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when getting the schema for ADD_SONG command should return the correct value 1`] = `
Object {
  "title": Object {
    "isOptional": true,
    "type": "NON_EMPTY_STRING",
  },
  "titleEnglish": Object {
    "isOptional": true,
    "type": "NON_EMPTY_STRING",
  },
}
`;
